src/schemes/fluid/implicit_incompressible_sph/system.jl

```diff
@@ -0,0 +1,563 @@
+
+    foreach_point_neighbor(system, neighbor_system,
+                           system_coords, neighbor_system_coords, semi;
+                           points=each_moving_particle(system)) do particle,
+                                                                   neighbor,
+                                                                   pos_diff,
+                                                                   distance
+        grad_kernel = smoothing_kernel_grad(system, pos_diff, distance, particle)
+
+        # Compute d_ji according to eq. 9 in Ihmsen et al. (2013).
+        # Note that we compute d_ji and not d_ij. We can use the antisymmetry
+        # of the kernel gradient and just flip the sign of W_ij to obtain W_ji.
+        d_ji_ = -time_step^2 * hydrodynamic_mass(system, particle) /
+                system.density[particle]^2 * (-grad_kernel)
+
+        d_ii_ = d_ii(system, particle)
+        m_b = hydrodynamic_mass(neighbor_system, neighbor)
+
+        # According to eq. 12 in Ihmsen et al. (2013)
+        a_ii[particle] += m_b * dot((d_ii_ - d_ji_), grad_kernel)
+    end
+end
+
+# Calculation of the contribution of the boundary particles the diagonal elements (a_ii-values)
+# according to Ihmsen et al. (2013)
+function calculate_diagonal_elements!(a_ii, system, neighbor_system::BoundarySystem, v, u, v_ode, u_ode, semi, time_step)
+    u_neighbor_system = wrap_u(u_ode, neighbor_system, semi)
+    system_coords = current_coordinates(u, system)
+    neighbor_system_coords = current_coordinates(u_neighbor_system, neighbor_system)
+
+    foreach_point_neighbor(system, neighbor_system,
+                           system_coords, neighbor_system_coords, semi;
+                           points=each_moving_particle(system)) do particle,
+                                                                   neighbor,
+                                                                   pos_diff,
+                                                                   distance
+        grad_kernel = smoothing_kernel_grad(system, pos_diff, distance, particle)
+
+        d_ii_ = d_ii(system, particle)
+        m_b = hydrodynamic_mass(neighbor_system, neighbor)
+
+        # Contribution to the diagonal elements without d_ji value (see eq. 16 in Ihmsen et al. (2013))
+        a_ii[particle] += m_b * dot(d_ii_, grad_kernel)
+    end
+end
+
+# Calculate pressure values with iterative pressure solver (relaxed Jacobi scheme)
+function pressure_solve(system, v, u, v_ode, u_ode, semi, t)
+    (; reference_density, max_error, min_iterations, max_iterations, time_step) = system
+
+    l = 1
+    terminate = false
+    # Convert relative error in percent to absolute error
+    eta = max_error * 0.01 * reference_density
+    while (!terminate)
+        @trixi_timeit timer() "pressure solver iteration" begin
+            avg_density_error = pressure_solve_iteration(system, avg_density_error, u,
```

Comment:
```suggestion
            avg_density_error = pressure_solve_iteration(system, u,
```
This fixes the dam break.